File: C/Data Structure/Doubly_LL.c
#include<stdio.h>
#include<stdlib.h>

struct node                 // size 12 byte
{
    int data ;              // size 4 bytes
    struct node *next ;     // size 4 bytes
    struct node *prev ;     // size 4 bytes
};

typedef struct node NODE ;
typedef struct node* PNODE ;
typedef struct node** PPNODE ;

void InsertFirst(PPNODE head , int no)
{
    PNODE newn = NULL ;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no ;
    newn->next = NULL ;
    newn->prev = NULL ;

    if(*head == NULL)
    {
        *head = newn ;
    }
    else
    {
        newn->next = *head ;        // 1st initilise newnode next 
        (*head)->prev = newn ;      // 2nd initilise firstnode prev
        *head = newn ;              // 3rd assigine newnode as first node
    }
}

void InsertLast(PPNODE head , int no)
{
    PNODE newn = NULL ;
    PNODE temp = NULL ;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = no ;
    newn->next = NULL ;
    newn->prev = NULL ;

    if(*head == NULL)
    {
        *head = newn ;
    }
    else
    {
        temp = *head ;

        while(temp->next != NULL)       // stop before last element
        {
            temp = temp->next ;         // go to last element
        }
        temp->next = newn ;            // insitile last node next value  
        newn->prev = temp ;            // insitile new node prev value
        
    }
}

void Display(PNODE head)
{
    printf("NULL <=> ");
    while(head != NULL)
    {
        printf("| %d |<=>",head->data);
        head = head->next ;
    }
    printf(" NULL \n");
}

int Count(PNODE head)
{
    int iCount = 0 ;

    while(head != NULL)
    {
        iCount++ ;
        head = head->next ;
    }
    return iCount ;
}

void DeleteFirst(PPNODE head)       // Delete first node without using temp pointer
{
    if(*head == NULL)
    {
        printf("LL is Allready empty !!!\n");
        return ;
    }
    else if((*head)->next == NULL)
    {
        free((*head)->next);
        *head = NULL ;
    }
    else
    {
        *head = (*head)->next ;
        free((*head)->prev);
        (*head)->prev = NULL ;
    }
}

void DeleteLast(PPNODE head)
{
    PNODE temp = NULL ;

    if((*head) == NULL)
    {
        printf("LL is Allready empty !!!\n");
        return ;
    }
    else if((*head)->next == NULL)
    {
        free(*head);
        *head = NULL ;
    }
    else
    {
        temp = *head ;

        while(temp->next->next != NULL)
        {
            temp = temp->next ;
        }
        free(temp->next);
        temp->next = NULL ;
    }
}

void InsertAtPos(PPNODE head , int no , int iPos)
{
    int iCountNode = 0 , iCnt = 0 ;

    PNODE newn = NULL ;
    PNODE temp = NULL ;

    iCountNode = Count(*head);

    if(iPos < 1 || iPos > iCountNode+1)
    {
        printf("Invalid Position !!! \n");
        return  ;
    }

    if(iPos == 1)
    {
        InsertFirst(head,no);
    }
    if(iPos == iCountNode+1)
    {
        InsertLast(head,no);
    }
    else
    {
        newn = (PNODE)malloc(sizeof(NODE));

        newn->data = no ;
        newn->next = NULL ;
        newn->prev = NULL ;

        temp = *head ;

        for(iCnt = 1 ; iCnt <= iPos-1 ; iCnt++);
        {
            temp = temp->next ;
        }
        newn->next = temp->next ;       // inisitilise newnode next 
        temp->next->prev = newn ;       // initilise newnode nextnode prev value
        temp->next = newn ;             // iniitiise newnode prevnode next value
        newn->prev = temp ;             // iniitilise newnode preves 
        
    }
}

void DeleteAtPos(PPNODE head , int iPos)
{
    int iCountNode = 0 , iCnt = 0  ;

    iCountNode = Count(*head);

    PNODE target = NULL ;
    PNODE temp = NULL ;

    if(iPos < 1 || iPos > iCountNode)
    {
        printf("Invalid Position !!! \n");
        return  ;
    }

    if(iPos == 1)
    {
        DeleteFirst(head);
    }
    if(iPos == iCountNode)
    {
        DeleteLast(head);
    }
    else
    {
        temp = *head ;

        for(iCnt = 1 ; iCnt < iPos-1 ; iCnt++)
        {
            temp = temp->next ;
        }
         
        target = temp->next ;           // store delete node into target variable
        temp->next = target->next ;     // insitilse next value of delete node before
        temp->next->prev = temp ;       // insitilse prev value of delete node After
        free(target);                   // Delete node   
    }
}

int main()
{
    PNODE first = NULL ;       // first is pointer

    int iOption = 0 , iNo = 0 , iPos = 0 , iRet = 0 ;

    while(1)
    {
        printf("-----------------------------------------------------\n");
        printf("1 : Insert the element at first Position \n");
        printf("2 : Insert the element at Last Position \n");
        printf("3 : Insert the element at Required Position \n");
        printf("4 : Delete the element from first Position \n");
        printf("5 : Delete the element from Last Position \n");
        printf("6 : Delete the element from Required Position \n");
        printf("7 : Display the Linked List \n");
        printf("8 : Total Nodes Counts of LL \n");
        printf("9 : Exit \n");
        printf("-----------------------------------------------------\n");
        printf("Enter the Option : \n");
        scanf("%d",&iOption);
        printf("-----------------------------------------------------\n");

        if(iOption==1)
        {
            printf("Enter the Data : \n");
            scanf("%d",&iNo);

            InsertFirst(&first,iNo);

            printf("Element succesfully inserted .\n");
        }
        else if(iOption == 2)
        {
            printf("Enter the Data : \n");
            scanf("%d",&iNo);

            InsertLast(&first,iNo);

            printf("Element succesfully inserted .\n");
        }
        else if(iOption == 3)
        {
            printf("Enter the Position : \n");
            scanf("%d",&iPos);

            printf("Enter the Data : \n");
            scanf("%d",&iNo);

            InsertAtPos(&first,iNo,iPos);

            printf("Element succesfully inserted .\n");
        }
        else if(iOption == 4)
        {
            DeleteFirst(&first);

            printf("First Element Deleted succesfully .\n");
        }
        else if(iOption == 5)
        {
            DeleteLast(&first);

            printf("Last Element Deleted succesfully .\n");
        }
        else if(iOption == 6)
        {
            printf("Enter the Position to delete element : \n");
            scanf("%d",&iPos);

            DeleteAtPos(&first,iPos);

            printf("Element Deleted succesfully .\n");
        }
        else if(iOption == 7)
        {
            Display(first);
        }
        else if(iOption == 8)
        {
            iRet = Count(first);
            printf("Total elements in LL are : %d\n",iRet);
        }
        else if(iOption == 9)
        {
            break;
        }
        else
        {
            printf("Invalid Option !!!! \n");
        }
    }
    return 0 ;
}

File: JAVA/Data Structure/Doubly_LL.java
import java.util.Scanner ;

class Node
{
    int data ;
    Node next ;
    Node prev ;

    Node(int iNo)
    {
        data = iNo ;
        next = null ;
        prev = null ;
    }
}

class Doubly
{
    // characterstis
    public Node head ;
    public Node tail ;
    public int iCountNode ;

    // Constructor
    public Doubly()
    {
        head = null ;
        tail = null ;
        iCountNode = 0 ;
    }

    // this method is used to count total nodes of LL
    public int countNode()
    {
        if((head == null) && (tail == null))
        {
            return 0 ;
        }

        return iCountNode ;
    }

    // this method is used to Display the LL
    public void displayLL()
    {
        if((head == null) && (tail == null))
        {
            System.out.println("LL is empty !!");
            return ;
        }

        Node temp = null ;

        temp = head ;

        System.out.print("null->");
        while(temp != null)
        {
            System.out.print("| "+ temp.data + " |<=>") ;
            temp = temp.next ;
        }
        System.out.print("null \n");
    }

    // this method is used to insert new node at first position of LL
    public void insertFirst(int iNo)
    {
        Node newn = null ;

        newn = new Node(iNo);

        if(head == null && tail == null)
        {
            head = newn ;
            tail = newn ;
        }
        else
        {
            newn.next = head ;
            head.prev = newn ;
            head = newn ;
        }

        iCountNode++ ;

        System.out.println("Data inserted succesfully.");
    }

    // this method is used to insert new node at last position of LL
    public void insertLast(int iNo)
    {
        Node newn = null ;

        newn = new Node(iNo);

        if(head == null && tail == null)
        {
            head = newn ;
            tail = newn ;
        }
        else
        {
            newn.prev = tail ;
            tail.next = newn ;
            tail = newn ;
        }
        iCountNode ++;

        System.out.println("Data inserted succesfully.");
    }

    // this method is used to delete the node from first position of LL
    public void deleteFirst()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is empty , unable to delete element !!");
            return ;
        }

        if(head == tail)
        {
            head = null ;
            tail = null ;

            System.gc();
        }
        else
        {
            head = head.next ;
            head.prev = null ;
            
            System.gc();
        }
        iCountNode-- ;

        System.out.println("Data deleted succesfully.");
    }

    // this method is used to delete node from last postion of LL
    public void deleteLast()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is empty , unable to delete element !!");
            return ;
        }

        if(head == tail)
        {
            head = null ;
            tail = null ;

            System.gc();
        }
        else
        {
            tail = tail.prev ;
            tail.next = null ;

            System.gc();
        }

        iCountNode-- ;

        System.out.println("Data deleted succesfully.");
    }

    // this method is used to insert new node at your required position of LL
    public void insertAtPos(int iNo , int iPos)
    {
        if(iPos < 1 || iPos > iCountNode+1)
        {
            System.out.println("invalid position !! , please enter correct position ");
            return ;
        }

        if(iPos == 1)
        {
            insertFirst(iNo);
        }
        else if(iPos == iCountNode+1)
        {
            insertLast(iNo);
        }
        else
        {
            Node newn = null ;
            Node temp = null ;
            int i = 0 ;

            newn = new Node(iNo);

            temp = head ;

            for(i = 1 ; i < iPos-1 ; i++)
            {
                temp = temp.next ;
            }
            newn.next = temp.next ;
            temp.next.prev = newn ;
            newn.prev = temp ;
            temp.next = newn ;

            iCountNode++ ;

            System.out.println("Data inserted succesfully.");
        }
    }

    // this method is used to delete the node from your required position of LL
    public void deleteAtPos(int iPos)
    {
        if(iPos < 1 || iPos > iCountNode)
        {
            System.out.println("invalid position !! , please enter correct position ");
            return ;
        }

        if(iPos == 1)
        {
            deleteFirst();
        }
        else if(iPos == iCountNode)
        {
            deleteLast();
        }
        else
        {
            Node temp = null ;
            int i = 0 ;

            temp = head ;

            for(i = 1 ; i < iPos-1 ; i++)
            {
                temp = temp.next ;
            }
            temp.next.next.prev = temp ;
            temp.next = temp.next.next ;

            iCountNode-- ;

            System.out.println("Data deleted succesfully.");
        }
    }

    // this method is used to delete all the node from LL and make LL empty
    public void emptyLL()
    {
        if(head == null && tail == null)
        {
            System.out.println("LL is already empty !!");
            return ;
        }

        Node temp = head ;
        Node nextNode = null ;

        while(temp != null)
        {
            nextNode = temp.next ;

            temp.next = null ;
            temp.prev = null ;

            temp = nextNode ;

            iCountNode-- ;
        }

        head = null ;
        tail = null ;

        System.out.println("Now , LL is empty");
    }
}

class Doubly_LL
{
    public static void main(String arg[])
    {
        Scanner scobj = new Scanner(System.in);

        int iOption = 0 , iRet = 0 , iData = 0 , iPos = 0 ;

        Doubly dobj = new Doubly();

        System.out.println("NOTE :\n1 : position of LL start with 1");
        while(true)
        {
            System.out.println("-----------------------------------------------------------------");
            System.out.println("0 : Terminate the Application");
            System.out.println("1 : Display the LL ");
            System.out.println("2 : Count the total nodes of LL");
            System.out.println("3 : Insert the data at first position");
            System.out.println("4 : Insert the data at last position");
            System.out.println("5 : Insert the data at required position");
            System.out.println("6 : Delete the data of first position");
            System.out.println("7 : Delete the data of last position");
            System.out.println("8 : Delete the data from requrired position");
            System.out.println("9 : Delete the all data from LL (OR) make LL empty");
            System.out.println("-----------------------------------------------------------------");
            System.out.println("Enter the option : ");
            iOption = scobj.nextInt();
            System.out.println("-----------------------------------------------------------------");
            
            if(iOption == 0)
            {
                System.out.println("Application is closed..");
                break ;
            }
            else if(iOption == 1)
            {
                dobj.displayLL();
            }
            else if(iOption == 2)
            {
                iRet = dobj.countNode();
                System.out.println("Total nodes in LL are : "+iRet);
            }
            else if(iOption == 3)
            {
                System.out.println("Enter the data : ");
                iData = scobj.nextInt();

                dobj.insertFirst(iData);
            }
            else if(iOption == 4)
            {
                System.out.println("Enter the data : ");
                iData = scobj.nextInt();

                dobj.insertLast(iData);
            }
            else if(iOption == 5)
            {
                System.out.println("Enter the data : ");
                iData = scobj.nextInt();

                System.out.println("Enter the position : ");
                iPos = scobj.nextInt();

                dobj.insertAtPos(iData,iPos);
            }
            else if(iOption == 6)
            {
                dobj.deleteFirst();
            }
            else if(iOption == 7)
            {
                dobj.deleteLast();
            }
            else if(iOption == 8)
            {
                System.out.println("Enter the position : ");
                iPos = scobj.nextInt();

                dobj.deleteAtPos(iPos);
            }
            else if(iOption == 9)
            {
                dobj.emptyLL();
            }
            else
            {
                System.out.println("!! Invalid option !!");
            }
        }
    }
}